src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Diagnostics/DocumentPullDiagnosticsEndpoint.cs

```diff
@@ -189,20 +190,36 @@ private async ValueTask ReportRZ10012TelemetryAsync(DocumentContext documentCont
             return;
         }
 
-        var document = await documentContext.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
-        var tagHelpers = document.GetTagHelpers();
+        var tagHelpers = await documentContext.Project.GetTagHelpersAsync(cancellationToken).ConfigureAwait(false);
+        var tagHelperCount = tagHelpers.Count();
+        var shouldReport = false;
 
-        if (Interlocked.Exchange(ref _lastReportedTagHelperCount, tagHelpers.Count) != tagHelpers.Count)
+        ImmutableInterlocked.AddOrUpdate(
+            ref _lastReporedProjectTagHelperCount,
+            documentContext.Project.Key,
+            (k) =>
+            {
+                shouldReport = true;
+                return tagHelperCount;
+            },
+            (k, currentValue) =>
+            {
+                shouldReport = currentValue != tagHelperCount;
+                return tagHelperCount;
+            });
+
+        if (shouldReport)
         {
             _telemetryReporter.ReportEvent(
                 "RZ10012",
                 Severity.Low,
-                new Property("tagHelpers", tagHelpers.Count),
-                new Property("RZ10012errors", relevantDiagnosticsCount));
+                new("tagHelpers", tagHelperCount),
+                new("RZ10012errors", relevantDiagnosticsCount),
+                new("project", documentContext.Project.Key.Id));
```

Comment:
Isn't this PII? Do have `PIIProperty`?

Reply:
It gets labeled in the server to not depend on client side labeling and ensure correct handling